src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,219 @@
+            if (EnsureDistributedService(context, out distributedCache))
+            {
+                return SetDistributedAsync(key, renderContent, context, distributedCache);
+            }
+            else
+            {
+                return SetMemoryAsync(key, renderContent, context);
+            }
+        }
+
+        private async Task<IHtmlContent> SetMemoryAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            var options = new MemoryCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            if (context.HasOptions && context.Options.ContainsKey(PriorityOptionName))
+            {
+                CacheItemPriority priority;
+                if (Enum.TryParse(context.Options[PriorityOptionName], out priority))
+                {
+                    options.SetPriority(priority);
+                }
+            }
+
+            // Create an entry link scope and flow it so that any tokens related to the cache entries
+            // created within this scope get copied to this scope.
+            using (var link = _memoryCache.CreateLinkingScope())
+            {
+                var content = await renderContent();
+
+                options.AddEntryLink(link);
+
+                var stringBuilder = new StringBuilder();
+                using (var writer = new StringWriter(stringBuilder))
+                {
+                    content.WriteTo(writer, context.HtmlEncoder);
+                }
+
+                var result = new StringBuilderHtmlContent(stringBuilder);
+
+                return _memoryCache.Set(key, result, options) as IHtmlContent;
+            }
+        }
+
+        public async Task<IHtmlContent> SetDistributedAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context, IDistributedCache cache)
+        {
+            var options = new DistributedCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            var content = await renderContent();
+
+            var stringBuilder = new StringBuilder();
+            using (var writer = new StringWriter(stringBuilder))
```

Comment:
Looks like this could you'd need something like `HttpResponseStreamWriter` to avoid creating large strings here. @rynowak, would you recommend using that type here?